{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "69ef182e-de91-4b83-8198-891140d5a3b0",
   "metadata": {},
   "source": [
    "# Spatial transcriptomics unveils the in situ cellular and molecular hallmarks of the lung in fatal COVID-19"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63baee74-cd69-4f24-a843-ade357a31684",
   "metadata": {},
   "source": [
    "# Pneumocytes specific expression of CAV1 in spatial coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2354c03-2564-4e7b-b841-82a8adc5e388",
   "metadata": {},
   "source": [
    "**Author:** Carlos A. Garcia-Prieto\n",
    "\n",
    "* This notebook explains how to plot cell type specific expression of genes in spatial coordinates using cell2location \n",
    "* We followed cell2location plotting cell-type specific expression of genes in spatial coordinates [tutorial.](https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_tutorial.html#Estimate-cell-type-specific-expression-of-every-gene-in-the-spatial-data-(needed-for-NCEM)) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b7b453d-050b-4b2a-941c-ea028d22d269",
   "metadata": {},
   "source": [
    "## Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d9c1a4bf-05cc-4001-9497-e7e7b086d409",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Global seed set to 0\n"
     ]
    }
   ],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import os\n",
    "import cell2location\n",
    "from cell2location.utils import select_slide\n",
    "import matplotlib as mpl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b78e243-4e52-4c68-a38b-62023ca10f52",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set data directory\n",
    "results_folder = \"/mnt/beegfs/cgarcia/Spatial/COVID19/cell2location/HLCA_publication/HLCA/\"\n",
    "ref_run_name = f'{results_folder}reference_signatures_finest'\n",
    "run_name = f'{results_folder}cell2location_map_finest'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9c57285-4cfc-4920-a858-5158d43aecbe",
   "metadata": {},
   "source": [
    "## Read cell type specific gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e5bc20c0-4b25-49fb-9e49-8a149dae77d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read cell type specific gene expression needed for NCEM\n",
    "adata_file = f\"{run_name}/ad_vis_post_distrib_finest_cell_type_gene_expr.h5ad\"\n",
    "adata_vis = sc.read_h5ad(adata_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15bfa70e-2583-4917-b23b-620fda6ee91a",
   "metadata": {},
   "source": [
    "# Plot cell type specific expression of CAV1 gene in spatial coordinates\n",
    "## Below we plot the cell type specific expression of CAV1 compared to total expression of the gene (first panel).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8814a262-4540-4e20-bca7-8ca2c78db492",
   "metadata": {},
   "source": [
    "Note that `plot_genes_per_cell_type` function often need customization so it is not included into cell2location package - you need to copy it from https://github.com/BayraktarLab/cell2location/blob/master/docs/notebooks/tutorial_utils.py to use on your system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fb24a9dc-353e-4b54-8487-97a5549940e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_genes_per_cell_type(slide, genes, ctypes):\n",
    "\n",
    "    n_genes = len(genes)\n",
    "    n_ctypes = len(ctypes)\n",
    "    fig, axs = plt.subplots(\n",
    "        nrows=n_genes, ncols=n_ctypes + 1, figsize=(4.5 * (n_ctypes + 1) + 2, 5 * n_genes + 1), squeeze=False\n",
    "    )\n",
    "    # axs = axs.reshape((n_genes, n_ctypes+1))\n",
    "\n",
    "    # plots of every gene\n",
    "    for j in range(n_genes):\n",
    "\n",
    "        # limit color scale at 99.9% quantile of gene expression (computed across cell types)\n",
    "        quantile_across_ct = np.array(\n",
    "            [\n",
    "                np.quantile(slide.layers[n][:, slide.var[\"SYMBOL\"] == genes[j]].toarray(), 0.999) \n",
    "                for n in slide.uns[\"mod\"][\"factor_names\"]\n",
    "            ]\n",
    "        )\n",
    "        quantile_across_ct = np.partition(quantile_across_ct.flatten(), -2)[-2]\n",
    "        sc.pl.spatial(\n",
    "            slide,\n",
    "            cmap=\"magma\",\n",
    "            color=genes[j],\n",
    "            # layer=ctypes[i],\n",
    "            gene_symbols=\"SYMBOL\",\n",
    "            ncols=4,\n",
    "            size=1.3,\n",
    "            img_key=\"hires\",\n",
    "            # limit color scale at 99.9% quantile of gene expression\n",
    "            vmin=0,\n",
    "            vmax=\"p99.9\",\n",
    "            ax=axs[j, 0],\n",
    "            show=False,\n",
    "        )\n",
    "\n",
    "        # plots of every cell type\n",
    "        for i in range(n_ctypes):\n",
    "            sc.pl.spatial(\n",
    "                slide,\n",
    "                cmap=\"magma\",\n",
    "                color=genes[j],\n",
    "                layer=ctypes[i],\n",
    "                gene_symbols=\"SYMBOL\",\n",
    "                ncols=4,\n",
    "                size=1.3,\n",
    "                img_key=\"hires\",\n",
    "                # limit color scale at 99.9% quantile of gene expression\n",
    "                vmin=0,\n",
    "                vmax=quantile_across_ct,\n",
    "                ax=axs[j, i + 1],\n",
    "                show=False,\n",
    "            )\n",
    "            axs[j, i + 1].set_title(f\"{genes[j]} {ctypes[i]}\")\n",
    "\n",
    "    return fig, axs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14165a67-1236-41d9-af6b-f2c1ac0f42ba",
   "metadata": {},
   "source": [
    "## Plot cell type specific gene expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "db1e4146-5302-4597-9a89-b60fe0821567",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create directory for Plot\n",
    "directory = f'{run_name}/Plots'\n",
    "# Check if the directory exists\n",
    "if not os.path.exists(directory):\n",
    "    # If it doesn't exist, create it\n",
    "    os.makedirs(directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e292448-4032-431d-a84b-288a7309dfa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select gene and cell types\n",
    "gene = ['CAV1'] \n",
    "ctypes = ['AT1', 'AT2', 'AT0']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "07c7cf87-6a9c-4066-9213-9297b58c60f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set sample used for spatio temporal trajectory analysis\n",
    "sample = \"HRC11\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5d32061d-ec38-4e95-9d44-f4efe548a191",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Select sample\n",
    "slide = select_slide(adata_vis, f'{sample}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "707acb52-24e5-485e-9246-3f85dba7626f",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\">\n",
    "<b>Paper Figure!</b>\n",
    "Panel Figure 5C\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "27a2802b-45ae-4eb3-bc4b-14bea1e90ec8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot CAV1 cell type specific expression\n",
    "with mpl.rc_context({'axes.facecolor':  'white', 'figure.figsize': [5, 5]}):\n",
    "    #from tutorial_utils import plot_genes_per_cell_type\n",
    "    plot_genes_per_cell_type(slide, gene, ctypes)\n",
    "    plt.savefig(f\"{run_name}/Plots/{sample}_pneumocytes_{gene}_gene_expression_Paper.png\",dpi=300, format=\"png\",pad_inches=0.2,bbox_inches=\"tight\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28b38fe1-d97d-406c-b8db-78c3e0692266",
   "metadata": {},
   "source": [
    "### Modules and their versions used for this analysis\n",
    "\n",
    "\n",
    "Useful for debugging and reporting issues."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a53fc38f-0639-43c4-9ea4-afe39c49bb72",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sys 3.9.16 (main, Jan 11 2023, 16:05:54) \n",
      "[GCC 11.2.0]\n",
      "re 2.2.1\n",
      "ipykernel._version 6.9.1\n",
      "json 2.0.9\n",
      "jupyter_client._version 7.2.2\n",
      "traitlets._version 5.1.1\n",
      "traitlets 5.1.1\n",
      "logging 0.5.1.2\n",
      "platform 1.0.8\n",
      "_ctypes 1.1.0\n",
      "ctypes 1.1.0\n",
      "zmq.sugar.version 23.2.0\n",
      "zmq.sugar 23.2.0\n",
      "zmq 23.2.0\n",
      "argparse 1.1\n",
      "tornado 6.1\n",
      "zlib 1.0\n",
      "colorama 0.4.6\n",
      "_curses b'2.2'\n",
      "dateutil._version 2.8.2\n",
      "dateutil 2.8.2\n",
      "six 1.16.0\n",
      "_decimal 1.70\n",
      "decimal 1.70\n",
      "jupyter_core.version 4.10.0\n",
      "jupyter_core 4.10.0\n",
      "entrypoints 0.4\n",
      "jupyter_client 7.2.2\n",
      "ipykernel 6.9.1\n",
      "IPython.core.release 8.4.0\n",
      "executing.version 0.8.3\n",
      "executing 0.8.3\n",
      "pure_eval.version 0.2.2\n",
      "pure_eval 0.2.2\n",
      "stack_data.version 0.2.0\n",
      "stack_data 0.2.0\n",
      "pygments 2.11.2\n",
      "ptyprocess 0.7.0\n",
      "pexpect 4.8.0\n",
      "IPython.core.crashhandler 8.4.0\n",
      "pickleshare 0.7.5\n",
      "backcall 0.2.0\n",
      "decorator 5.1.1\n",
      "_sqlite3 2.6.0\n",
      "sqlite3.dbapi2 2.6.0\n",
      "sqlite3 2.6.0\n",
      "wcwidth 0.2.5\n",
      "prompt_toolkit 3.0.20\n",
      "parso 0.8.3\n",
      "jedi 0.18.1\n",
      "urllib.request 3.9\n",
      "IPython.core.magics.code 8.4.0\n",
      "IPython 8.4.0\n",
      "setuptools._distutils 3.9.16\n",
      "pkg_resources._vendor.more_itertools 8.12.0\n",
      "pkg_resources.extern.more_itertools 8.12.0\n",
      "pkg_resources._vendor.appdirs 1.4.3\n",
      "pkg_resources.extern.appdirs 1.4.3\n",
      "pkg_resources._vendor.packaging.__about__ 21.3\n",
      "pkg_resources._vendor.packaging 21.3\n",
      "pkg_resources.extern.packaging 21.3\n",
      "pkg_resources._vendor.pyparsing 3.0.9\n",
      "pkg_resources.extern.pyparsing 3.0.9\n",
      "setuptools.version 65.6.3\n",
      "setuptools._vendor.packaging.__about__ 21.3\n",
      "setuptools._vendor.packaging 21.3\n",
      "setuptools.extern.packaging 21.3\n",
      "setuptools._vendor.ordered_set 3.1\n",
      "setuptools.extern.ordered_set 3.1\n",
      "setuptools._vendor.more_itertools 8.8.0\n",
      "setuptools.extern.more_itertools 8.8.0\n",
      "_csv 1.0\n",
      "csv 1.0\n",
      "setuptools._vendor.pyparsing 3.0.9\n",
      "setuptools.extern.pyparsing 3.0.9\n",
      "setuptools 65.6.3\n",
      "distutils 3.9.16\n",
      "debugpy 1.5.1\n",
      "xmlrpc.client 3.9\n",
      "socketserver 0.4\n",
      "http.server 0.6\n",
      "_pydevd_frame_eval.vendored.bytecode 0.13.0.dev\n",
      "_pydev_bundle.fsnotify 0.1.5\n",
      "pydevd 2.6.0\n",
      "numpy.version 1.22.4\n",
      "numpy.core._multiarray_umath 3.1\n",
      "numpy.core 1.22.4\n",
      "numpy.linalg._umath_linalg 0.1.5\n",
      "numpy.lib 1.22.4\n",
      "numpy 1.22.4\n",
      "packaging 23.0\n",
      "matplotlib._version 3.6.3\n",
      "PIL._version 9.4.0\n",
      "PIL 9.4.0\n",
      "PIL._deprecate 9.4.0\n",
      "PIL.Image 9.4.0\n",
      "pyparsing 3.0.9\n",
      "cycler 0.10.0\n",
      "kiwisolver._cext 1.4.4\n",
      "kiwisolver 1.4.4\n",
      "matplotlib 3.6.3\n",
      "scanpy._metadata 1.9.1\n",
      "scipy.version 1.10.0\n",
      "scipy 1.10.0\n",
      "scipy.sparse.linalg._isolve._iterative b'$Revision: $'\n",
      "scipy._lib.decorator 4.0.5\n",
      "scipy.linalg._fblas b'$Revision: $'\n",
      "scipy.linalg._flapack b'$Revision: $'\n",
      "scipy.linalg._flinalg b'$Revision: $'\n",
      "scipy.sparse.linalg._eigen.arpack._arpack b'$Revision: $'\n",
      "anndata._metadata 0.8.0\n",
      "h5py.version 3.8.0\n",
      "h5py 3.8.0\n",
      "natsort 8.2.0\n",
      "pytz 2022.7.1\n",
      "tarfile 0.9.0\n",
      "pandas 1.5.3\n",
      "numcodecs.version 0.11.0\n",
      "numcodecs.blosc 1.21.0\n",
      "numcodecs.zstd 1.4.8\n",
      "numcodecs.lz4 1.9.3\n",
      "msgpack 1.0.4\n",
      "numcodecs 0.11.0\n",
      "fasteners 0.18\n",
      "zarr.version 2.13.6\n",
      "zarr 2.13.6\n",
      "yaml 5.4.1\n",
      "toolz 0.12.0\n",
      "tlz 0.12.0\n",
      "cloudpickle 2.2.1\n",
      "markupsafe 2.1.2\n",
      "dask 2023.1.1\n",
      "llvmlite 0.43.0\n",
      "numba.cloudpickle 3.0.0\n",
      "numba.misc.appdirs 1.4.1\n",
      "numba 0.60.0\n",
      "sparse 0.13.0\n",
      "scipy._lib._uarray 0.8.8.dev0+aa94c5a4.scipy\n",
      "scipy.special._specfun b'$Revision: $'\n",
      "anndata 0.8.0\n",
      "joblib.externals.cloudpickle 2.2.0\n",
      "joblib.externals.loky 3.3.0\n",
      "joblib 1.2.0\n",
      "sklearn.utils._joblib 1.2.0\n",
      "scipy.optimize._minpack2 b'$Revision: $'\n",
      "scipy.optimize._lbfgsb b'$Revision: $'\n",
      "scipy.optimize._cobyla b'$Revision: $'\n",
      "scipy.optimize._slsqp b'$Revision: $'\n",
      "scipy.optimize.__nnls b'$Revision: $'\n",
      "scipy.linalg._interpolative b'$Revision: $'\n",
      "scipy.integrate._vode b'$Revision: $'\n",
      "scipy.integrate._dop b'$Revision: $'\n",
      "scipy.integrate._lsoda b'$Revision: $'\n",
      "scipy.interpolate.dfitpack b'$Revision: $'\n",
      "scipy.stats._statlib b'$Revision: $'\n",
      "scipy.stats._mvn b'$Revision: $'\n",
      "threadpoolctl 3.1.0\n",
      "sklearn.base 1.2.1\n",
      "sklearn.utils._show_versions 1.2.1\n",
      "sklearn 1.2.1\n",
      "texttable 1.6.7\n",
      "igraph.version 0.9.11\n",
      "igraph 0.9.11\n",
      "leidenalg.version 0.8.10\n",
      "leidenalg 0.8.10\n",
      "louvain 0.7.2\n",
      "scanpy 1.9.1\n",
      "torch.version 1.13.1+cu117\n",
      "torch.torch_version 1.13.1+cu117\n",
      "opt_einsum v3.3.0\n",
      "torch.cuda.nccl (2, 14, 3)\n",
      "astunparse 1.6.3\n",
      "torch.backends.cudnn 8500\n",
      "tqdm._dist_ver 4.66.4\n",
      "tqdm.version 4.66.4\n",
      "tqdm.cli 4.66.4\n",
      "tqdm 4.66.4\n",
      "torch 1.13.1+cu117\n",
      "pyro._version 1.8.4\n",
      "pyro 1.8.4\n",
      "attr 22.2.0\n",
      "pytorch_lightning.__version__ 1.9.0\n",
      "lightning_fabric.__version__ 1.9.0\n",
      "lightning_utilities.__about__ 0.6.0.post0\n",
      "lightning_utilities 0.6.0.post0\n",
      "fsspec 2023.1.0\n",
      "lightning_fabric 1.9.0\n",
      "torchmetrics.__about__ 0.11.1\n",
      "torchmetrics 0.11.1\n",
      "ipaddress 1.0\n",
      "pytorch_lightning 1.9.0\n",
      "jaxlib.version 0.4.3\n",
      "jaxlib 0.4.3\n",
      "jax.version 0.4.3\n",
      "jax.lib 0.4.3\n",
      "etils 1.0.0\n",
      "jax 0.4.3\n",
      "wrapt 1.14.1\n",
      "tree 0.1.8\n",
      "xml.sax.handler 2.0beta\n",
      "chex 0.1.6\n",
      "mudata 0.2.1\n",
      "docrep 0.3.2\n",
      "urllib3.packages.six 1.16.0\n",
      "urllib3._version 1.26.14\n",
      "urllib3.util.ssl_match_hostname 3.5.0.1\n",
      "urllib3.connection 1.26.14\n",
      "urllib3 1.26.14\n",
      "charset_normalizer.version 2.1.1\n",
      "charset_normalizer 2.1.1\n",
      "requests.packages.urllib3.packages.six 1.16.0\n",
      "requests.packages.urllib3._version 1.26.14\n",
      "requests.packages.urllib3.util.ssl_match_hostname 3.5.0.1\n",
      "requests.packages.urllib3.connection 1.26.14\n",
      "requests.packages.urllib3 1.26.14\n",
      "idna.package_data 3.4\n",
      "idna.idnadata 15.0.0\n",
      "idna 3.4\n",
      "requests.packages.idna.package_data 3.4\n",
      "requests.packages.idna.idnadata 15.0.0\n",
      "requests.packages.idna 3.4\n",
      "certifi 2022.06.15\n",
      "requests.__version__ 2.28.2\n",
      "requests.utils 2.28.2\n",
      "requests 2.28.2\n",
      "flax.version 0.6.4\n",
      "flax 0.6.4\n",
      "optax 0.1.4\n",
      "multipledispatch 0.6.0\n",
      "numpyro.version 0.11.0\n",
      "numpyro 0.11.0\n",
      "scvi 0.20.0\n",
      "pynndescent 0.5.8\n",
      "google.protobuf 3.19.6\n",
      "tensorflow.python.client.pywrap_tf_session 2.11.0\n",
      "tensorflow.python.framework.versions 2.11.0\n",
      "tensorflow.python.keras 2.6.0\n",
      "flatbuffers._version 23.1.21\n",
      "flatbuffers 23.1.21\n",
      "tensorflow._api.v2.compat.v1.version 2.11.0\n",
      "tensorflow._api.v2.compat.v1.compat.v1 2.11.0\n",
      "tensorflow._api.v2.compat.v2.compat.v1 2.11.0\n",
      "tensorflow._api.v2.compat.v2.version 2.11.0\n",
      "tensorboard.version 2.11.2\n",
      "tensorboard 2.11.2\n",
      "tensorboard.compat.tensorflow_stub.pywrap_tensorflow 0\n",
      "tensorboard.compat.tensorflow_stub stub\n",
      "tensorflow._api.v2.compat.v2.compat.v2 2.11.0\n",
      "tensorflow._api.v2.compat.v2 2.11.0\n",
      "tensorflow._api.v2.compat.v1.compat.v2 2.11.0\n",
      "tensorflow._api.v2.compat.v1 2.11.0\n",
      "tensorflow._api.v2.version 2.11.0\n",
      "tensorflow.compat.v1.compat.v1 2.11.0\n",
      "tensorflow.compat.v1.compat.v2 2.11.0\n",
      "tensorflow.compat.v1.version 2.11.0\n",
      "tensorflow.compat.v1 2.11.0\n",
      "tensorflow.compat.v2.compat.v1 2.11.0\n",
      "tensorflow.compat.v2.compat.v2 2.11.0\n",
      "tensorflow.compat.v2.version 2.11.0\n",
      "tensorflow.compat.v2 2.11.0\n",
      "keras 2.11.0\n",
      "keras.api.keras 2.11.0\n",
      "keras.api._v2.keras 2.11.0\n",
      "tensorflow 2.11.0\n",
      "umap 0.5.3\n",
      "cell2location 0.1.3\n"
     ]
    }
   ],
   "source": [
    "cell2location.utils.list_imported_modules()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Environment (c2l_ncem_env)",
   "language": "python",
   "name": "c2l_ncem"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
